### PR TITLE
REF: Store metadata in an attrs dict

### PR DESCRIPTION
**Pull request by TomAugspurger**
This aids in the implementation of
https://github.com/pandas-dev/pandas/pull/28394. Over there, I'm having
issues with using `NDFrame.__finalize__` to copy attributes, in part
because getattribute on NDFrame is so complicated.

This simplifies that PR because we only need to look in NDFrame.attrs (name borrowed from xarray), 
which is just a plain dictionary.

Aside from the addition of a public NDFrame.attrs dictionary, there
aren't any user-facing API changes.

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -2029,7 +2049,13 @@ def to_dense(self):
 
     def __getstate__(self):
         meta = {k: getattr(self, k, None) for k in self._metadata}
-        return dict(_data=self._data, _typ=self._typ, _metadata=self._metadata, **meta)
+        return dict(
+            _data=self._data,
+            _typ=self._typ,
+            _metadata=self._metadata,
```
Comment (jbrockmendel) [inline pandas/core/generic.py]: are attrs linked to _metadata in any way?

Comment (TomAugspurger) [reply pandas/core/generic.py]: Not yet, though perhaps eventually if we can find a suitable way to deprecate `_metadata` for subclasses. With this PR, pandas doesn't use `_metadata` itself anymore (previously, we just used it for `Series.name`.

File: pandas/core/series.py [hunk 2]
```diff
@@ -485,6 +471,16 @@ def dtypes(self):
         """
         return self._data.dtype
 
+    @property
+    def name(self) -> Hashable:
```
Comment (TomAugspurger) [inline pandas/core/series.py]: @WillAyd @simonjayhawkins do you have a preferred type on Series.name? Previously I had `Optional[Hashable]`, but None is Hashable so either works.

In DataFrameGroupBy._iterate_slices we have its type as just `Hashable`.

Comment (WillAyd) [reply pandas/core/series.py]: I think this is fine but no strong preference

File: pandas/core/generic.py [hunk 3]
```diff
@@ -188,6 +190,9 @@ class NDFrame(PandasObject, SelectionMixin):
     _is_copy = None
     _data = None  # type: BlockManager
 
+    if TYPE_CHECKING:
```
Comment (WillAyd) [inline pandas/core/generic.py]: Why is this required in the TYPE_CHECKING block?

File: pandas/core/generic.py [hunk 4]
```diff
@@ -197,6 +202,7 @@ def __init__(
         axes: Optional[List[Index]] = None,
         copy: bool = False,
         dtype: Optional[Dtype] = None,
+        attrs: Optional[Mapping[Hashable, Any]] = None,
```
Comment (WillAyd) [inline pandas/core/generic.py]: Minor but any reason to prefer `Mapping` over `Dict`? I think should stick with the latter generically

File: pandas/core/generic.py [hunk 5]
```diff
@@ -233,6 +244,19 @@ def _init_mgr(self, mgr, axes=None, dtype=None, copy=False):
 
     # ----------------------------------------------------------------------
 
+    @property
+    def attrs(self) -> Dict[Hashable, Any]:
```
Comment (WillAyd) [inline pandas/core/generic.py]: Been a while since we've added something to a frame - do we have to make any consideration for dot access for frames where this was previously a valid label?

File: pandas/core/generic.py (hunk 3 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: Otherwise you'd have a mutable default shared by all the classes. Not sure if there's a better way, but this is the workaround I saw on some mypy thread.

File: pandas/core/generic.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: This is what xarray does. It's probably best to be forgiving in what we accept. This isn't user-facing (yet), but they could give an OrderedDict say.

File: pandas/core/generic.py (hunk 5 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: Don't think so. We have a blanket statement that `.` is a convenience.

File: pandas/core/generic.py (hunk 4 above)
Comment (WillAyd) [reply pandas/core/generic.py]: Dict would encompass OrderedDict since its its parent, so won't make a difference there. There are some edge cases where this would make a difference (i.e. if we called `attrs.copy()` but I think should just stick with `Dict` since we don't have this elsewhere and maybe leave to a separate exercise if we wanted to really fine tune those types

File: pandas/core/generic.py (hunk 3 above)
Comment (WillAyd) [reply pandas/core/generic.py]: Hmm can you maybe share the error? We already have the `_metadata` defaulting to `[]` so not clear on the distinction.

Comment (TomAugspurger) [reply pandas/core/generic.py]: There wouldn't be an error, it'd typecheck just fine.

It's a runtime issue though, since you could do `pd.Series.attrs['a'] = 'foo'`, which would affect all the instances.

With python 3.6+, we can just do

```
attrs : Dict[Hashable, Any]
```

but the workaround for 3.5 is to put it in a TYPE_CHECKING block.

Comment (WillAyd) [reply pandas/core/generic.py]: Ah OK. I wasn't getting the link between runtime and TYPE_CHECKING but understood now. So I guess should either wait for Py35 to drop or should add a TODO here to clean up when that happens and add the definition without declaration

File: pandas/core/generic.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: I'd prefer to match xarray here.

Comment (WillAyd) [reply pandas/core/generic.py]: I'm not against it but this isn't the place to start. We don't do this anywhere else.

```sh
$ grep "Dict\[" -r pandas --include="*.py" | wc -l
      60
$ grep "Mapping\[" -r pandas --include="*.py" | wc -l
       0
```

So if you feel strongly about matching xarray should be done in a separate PR comprehensively instead of just one-off here

Comment (TomAugspurger) [reply pandas/core/generic.py]: Maybe let's step back: why isn't Mapping the correct type?

Comment (simonjayhawkins) [reply pandas/core/generic.py]: from https://github.com/python/typeshed/blob/master/CONTRIBUTING.md

Some further tips for good type hints:

avoid invariant collection types (List, Dict) in argument positions, in favor of covariant types like Mapping or Sequence;

Comment (TomAugspurger) [reply pandas/core/generic.py]: The storage is a dict, the argument is a mapping.

Comment (WillAyd) [reply pandas/core/generic.py]: OK. I wasn't expecting this be such a sticking point so whatever everyone else is fine with works

@simonjayhawkins thanks for the link. Interestingly enough there are still more Dict appearances in typeshed than Mapping so guess it doesn't matter too much

File: pandas/core/generic.py (hunk 1 above)
Comment (jreback) [reply pandas/core/generic.py]: wait, then we *should* deprecate _metadata right?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: Most likely? I have no idea who is using it or what they're using it for though.

File: pandas/core/generic.py (hunk 5 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: FWIW, we don't need to expose this publicly yet. We can call it `._attrs` until we're ready.

My main motivation is ease of implementation in https://github.com/pandas-dev/pandas/pull/28394.

But xarray is an example of this being useful in practice.

File: pandas/core/generic.py (hunk 1 above)
Comment (jreback) [reply pandas/core/generic.py]: ok can you open an issue to deprecate that

File: pandas/core/generic.py [hunk 6]
```diff
@@ -5213,6 +5245,9 @@ def __finalize__(self, other, method=None, **kwargs):
 
         """
         if isinstance(other, NDFrame):
+            for name in other.attrs:
+                self.attrs[name] = other.attrs[name]
```
Comment (jreback) [inline pandas/core/generic.py]: self.attrs.update(other) ?

**Review by jreback — APPROVED** [review]
lgtm.

**Comment by jreback** [comment]
do you want to add docs anywhere / whatsnew? or just keep private mention for now?

File: pandas/core/generic.py (hunk 6 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: That'd work for now, but we'll need a for loop in my other PR since we'll have per-attr logic to resolve these.

**Comment by TomAugspurger** [comment]
> do you want to add docs anywhere / whatsnew? or just keep private mention for now?

I pushed some basics API docs. Will want a larger section once things settle I think.

**Comment by jreback** [comment]
thanks @TomAugspurger very nice!

File: pandas/core/series.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/series.py]: just ran into problems with this in `Series._unpickle_series_compat`, 

`pandas\core\series.py:1018:25: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "Hashable")  [assignment]`

so will need to change. see https://github.com/pandas-dev/pandas/pull/28339/commits/991e90d50d4c0a4a3ad4ff699e208a263dd76515

```python
from typing import Optional, Hashable

name: Optional[Hashable]
name2: Hashable

name = None
name2 = None  # error: Incompatible types in assignment (expression has type "None", variable has type "Hashable")  [assignment]
```

**Comment by shoyer** [comment]
In xarray, we currently use duck-typing to check for `.attrs` attributes in the `xarray.DataArray` constructor. So this PR broke our tests (see https://github.com/pydata/xarray/issues/3440), but that's probably our own fault.

My bigger concern is that I noticed that `attrs` is used in different way in pandas than xarray: `name` is a member of `attrs` vs. in xarray where `DataArray.name` is a separate attribute. This is not an unreasonable thing to do, but it does introduce some minor incompatible: in xarray (and h5py) `attrs` really is an arbitrary mapping, but that isn't quite true in pandas (because  `name` needs to be hashable). Also, what happens if you try to delete `name` from the `attrs` dict?

**Comment by jorisvandenbossche** [comment]
> Also, what happens if you try to delete name from the attrs dict?

From a quick test, it seems this deletes the name (`s.name` becomes `None`); which I *suppose* is the expected behaviour?

But you can indeed set it to a non-hashable value that way, which doesn't directly raise an error, but can confusingly give errors in later operations:

```

In [9]: s = pd.Series([1, 2, 3], name='a')                                                                                                                                                                         

In [10]: s.attrs                                                                                                                                                                                                   
Out[10]: {'name': 'a'}

In [11]: s.attrs = [1, 2]                                                                                                                                                                                          
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-d0401d6f5710> in <module>
----> 1 s.attrs = [1, 2]

~/scipy/pandas/pandas/core/generic.py in __setattr__(self, name, value)
   5252         try:
   5253             object.__getattribute__(self, name)
-> 5254             return object.__setattr__(self, name, value)
   5255         except AttributeError:
   5256             pass

~/scipy/pandas/pandas/core/generic.py in attrs(self, value)
    243     @attrs.setter
    244     def attrs(self, value: Mapping[Optional[Hashable], Any]) -> None:
--> 245         self._attrs = dict(value)
    246 
    247     def _validate_dtype(self, dtype):

TypeError: cannot convert dictionary update sequence element #0 to a sequence

In [12]: s.attrs['name'] = [1, 2]                                                                                                                                                                                  

In [13]: s                                                                                                                                                                                                         
Out[13]: 
0    1
1    2
2    3
Name: [1, 2], dtype: int64

In [14]: s.to_frame()                                                                                                                                                                                              
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-14-78ea8628cc3c> in <module>
----> 1 s.to_frame()

~/scipy/pandas/pandas/core/series.py in to_frame(self, name)
   1533         """
   1534         if name is None:
-> 1535             df = self._constructor_expanddim(self)
   1536         else:
   1537             df = self._constructor_expanddim({name: self})

~/scipy/pandas/pandas/core/frame.py in __init__(self, data, index, columns, dtype, copy)
    456                 data_columns = list(data.dtype.names)
    457                 data = {k: data[k] for k in data_columns}
--> 458                 if columns is None:
    459                     columns = data_columns
    460                 mgr = init_dict(data, index, columns, dtype=dtype)

TypeError: unhashable type: 'list'
```

**Comment by jorisvandenbossche** [comment]
Do we need to further discuss @shoyer's concern?

I was planning to try this out in GeoPandas (to see if it can replace our usage of `_metadata`), to be able to give feedback on this feature that way, but didn't get to it yet.

**Comment by TomAugspurger** [comment]
Sorry I missed @shoyer's comment. I'll take a look.

**Comment by TomAugspurger** [comment]
Pandas will need to call __finalize__ in more places to ensure the metadata is propagated.

**Comment by jorisvandenbossche** [comment]
@buhrmann can you open a new issue with this specific use case?

**Comment by dotsdl** [comment]
Is it appearing likely that `DataFrame.attrs` is here to stay? I'm considering using this for metadata used to differentiate DataFrames of two forms in [alchemlyb](https://alchemlyb.readthedocs.io/en/latest/parsing.html#standard-forms-of-raw-data), but wanted to check the temperature of the pandas devs on this question before relying on it. So far it works as expected (e.g. it survives such things as `DataFrame.copy()`)!
